falcon/cyutil/streams.pyx

```diff
@@ -0,0 +1,312 @@
+from libc.stdint cimport uint32_t
+
+import functools
+import io
+
+DEFAULT_CHUNK_SIZE = io.DEFAULT_BUFFER_SIZE * 4
+
+
+cdef class BufferedStream:
+
+    cdef _read_func
+    cdef Py_ssize_t _chunk_size
+
+    cdef bytes _buffer
+    cdef Py_ssize_t _buffer_len
+    cdef Py_ssize_t _buffer_pos
+    cdef Py_ssize_t _max_bytes_remaining
+
+    def __cinit__(self, read, max_stream_len, chunk_size=None):
+        self._read_func = read
+        self._chunk_size = chunk_size or DEFAULT_CHUNK_SIZE
+
+        self._buffer = b''
+        self._buffer_len = 0
+        self._buffer_pos = 0
+        self._max_bytes_remaining = max_stream_len
+
+    def peek(self, Py_ssize_t amount=-1):
+        cdef Py_ssize_t read_amount
+
+        amount = int(amount)
+        if amount < 0 or amount > self._chunk_size:
+            amount = self._chunk_size
+
+        if self._buffer_len - self._buffer_pos < amount:
+            read_amount = (self._chunk_size - self._buffer_len +
+                           self._buffer_pos)
+            if read_amount >= self._max_bytes_remaining:
```

Comment:
@vytas7 Hi, an amazing PR!

I have one more question, maybe in the future would be better to split `if/if/else` blocks to 3 different private methods of an instance?

e.g
```
if read_amount >= self._max_bytes_remaining:
    self._some_cool_method()
if self._buffer_pos == 0:
   self._some_cool_method2()
else:
   self.some_cool_method3()
```

What do you think?

Reply:
@vytas7 Understood, in that case it make sense.